pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+
+            // Only source or target can close a stream.
+            ensure!(
+                origin == stream.source || origin == stream.target,
+                Error::<T>::UnauthorizedOrigin
+            );
+
+            // Update stream before closing it to ensure fair payment.
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+
+            // Unfreeze funds left in the stream.
+            T::Assets::decrease_deposit(stream.asset_id.clone(), &stream.source, stream.deposit)?;
+
+            // Remove stream from storage.
+            Streams::<T>::remove(stream_id);
+            LookupStreamsWithSource::<T>::remove(stream.source, stream_id);
+            LookupStreamsWithTarget::<T>::remove(stream.target, stream_id);
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamClosed {
+                stream_id,
+                refunded: stream.deposit,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(2)]
+        pub fn update_stream(
+            origin: OriginFor<T>,
+            stream_id: T::StreamId,
+        ) -> DispatchResultWithPostInfo {
+            // No problem with anyone updating any stream.
+            let _ = ensure_signed(origin)?;
+
+            let mut stream = Streams::<T>::get(stream_id).ok_or(Error::<T>::UnknownStreamId)?;
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+            Streams::<T>::insert(stream_id, stream);
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(3)]
+        pub fn refill_stream(
+            origin: OriginFor<T>,
+            stream_id: T::StreamId,
+            increase: T::Balance,
+        ) -> DispatchResultWithPostInfo {
+            let origin = ensure_signed(origin)?;
+            let mut stream = Streams::<T>::get(stream_id).ok_or(Error::<T>::UnknownStreamId)?;
+
+            // Only source can refill stream
+            ensure!(origin == stream.source, Error::<T>::UnauthorizedOrigin);
+
+            // Source will not pay for drained stream retroactively, so we perform payment with
+            // what is left first.
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+
+            // Increase deposit.
+            T::Assets::increase_deposit(stream.asset_id.clone(), &origin, increase)?;
+            stream.deposit = stream
+                .deposit
+                .checked_add(&increase)
+                .ok_or(Error::<T>::CurrencyOverflow)?;
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamRefilled {
+                stream_id,
+                increase,
+                new_deposit: stream.deposit,
+            });
+
+            // Update stream info in storage.
+            Streams::<T>::insert(stream_id, stream);
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(4)]
+        pub fn change_stream_rate(
```

Comment:
Maybe it would also make sense to allow changing the time unit too? what do you think?